File: test/Interpreter/Inputs/objc_direct.m
#import "objc_direct.h"

@implementation Bar
- (instancetype)initWithValue:(int)value {
  printf("called %s with %d\n", __FUNCTION__, value);
  self = [super init];
  _directProperty = value;
  return self;
}
- (int)objectAtIndexedSubscript:(int)i {
  return 789;
}
- (void)setObject:(int)obj atIndexedSubscript:(int)i {}
- (NSString *)directMethod {
  return @"called directMethod";
}
+ (NSString *)directClassMethod {
  return @"called directClassMethod";
}
- (NSString *)directProtocolMethod {
  return @"called directProtocolMethod";
}
@end

@implementation Bar(CategoryName)
- (int)directProperty2 {
  return 456;
}
- (void)setDirectProperty2:(int)i {}
- (NSString *)directMethod2 {
  return @"called directMethod2";
}
+ (NSString *)directClassMethod2 {
  return @"called directClassMethod2";
}
@end

File: test/Interpreter/objc_direct.swift
// RUN: %empty-directory(%t)
// RUN: %target-clang %S/Inputs/objc_direct.m -I %S/../Inputs -c -o %t/objc_direct.o
// RUN: %target-build-swift -import-objc-header %S/../Inputs/objc_direct.h -Xlinker %t/objc_direct.o %s -o %t/a.out
// RUN: %target-codesign %t/a.out
// RUN: %target-run %t/a.out | %FileCheck %s

// REQUIRES: executable_test
// REQUIRES: objc_interop

protocol BarProtocol {
    func directProtocolMethod() -> String!
}

extension Bar: BarProtocol {}

let bar = Bar()

bar.directProperty = 123
print(bar.directProperty)
// CHECK: 123

bar.directProperty2 = 456
print(bar.directProperty2)
// CHECK: 456

bar[0] = 0
print(bar[0])
// CHECK: 789

print(bar.directMethod())
// CHECK: called directMethod

print(bar.directMethod2())
// CHECK: called directMethod2

print(Bar.directClassMethod())
// CHECK: called directClassMethod

print(Bar.directClassMethod2())
// CHECK: called directClassMethod2

print(bar.directProtocolMethod())
// CHECK: called directProtocolMethod
